### ISSUE TITLE
Allow specifying the "accent color" of a form control element

### ISSUE THREAD

**Issue by chrishtr** [w3c/csswg-drafts#5187]
(not sure which spec this might apply to - css-color-adjust? css-color? css-pseudo?)

Form controls often have a themed highlight color that indicates a "selected" state. For example, in Chromium, we recently changed the look and feel to have a blue highlight color:

<img width="271" alt="Screen Shot 2020-06-06 at 9 11 56 AM" src="https://user-images.githubusercontent.com/3453258/84091874-a4a48000-a9aa-11ea-94b7-5788767ed3e5.png">
<img width="217" alt="Screen Shot 2020-06-06 at 9 12 04 AM" src="https://user-images.githubusercontent.com/3453258/84091877-a706da00-a9aa-11ea-90ad-9c58710247a9.png">
<img width="187" alt="Screen Shot 2020-06-06 at 9 11 47 AM" src="https://user-images.githubusercontent.com/3453258/84091880-a9693400-a9aa-11ea-8dc0-4a2ee8ddbb96.png">
<img width="131" alt="Screen Shot 2020-06-06 at 9 11 42 AM" src="https://user-images.githubusercontent.com/3453258/84091886-abcb8e00-a9aa-11ea-9176-12452788ca65.png">

(Edge on Mac has similar styling to Chrome but does not use blue). [This demo](http://concrete-hardboard.glitch.me/) enumerates all of the form controls and can be tested on various browsers.

In addition, some operating systems, such as Mac OS X, allow changing the accent color via a user setting. Safari uses this setting to change the blue color in the screenshots above to a different color.

A number of developers have indicated that they would prefer to be able to change the color. See e.g. [this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1091531). I propose a way to do this via CSS:

* A new css property `accent-color` that would take on the usual color values. We would specify that a user agent *should* use this color to change the accent color of a UA-rendered form control, if there is such a well-defined concept for that form control; and otherwise this property has no effect. In the various accessibility and forced color modes, this property is ignored.

In general, form control styling and rendering is not specified, but this particular part of it seems like it can be.

**Comment by AmeliaBR** [comment]
Is there a reason this can't be done via `color` alone? Would that currently break the modern OS appearance of the control?

**Comment by chrishtr** [comment]
> Is there a reason this can't be done via `color` alone? Would that currently break the modern OS appearance of the control?

Two reasons: 
* Possibly breaking existing sites
* `color` doesn't indicate what is being colored. For example, there are non-background strokes like borders or the fill color of a progress input, does `color` change those?

**Comment by fantasai** [comment]
Accent color isn't indicating what's being colored, either. I'd say if there isn't a compat restriction, re-using `color` here makes a fair amount of sense unless there's a need to indicate two different colors as input.

FWIW, the CSS UI spec [used to say](https://www.w3.org/TR/2004/CR-css3-ui-20040511/#appearance1)
>  If 'appearance' is not 'normal', the UA must render the element as if it was the specified user interface (UI) control from the platform. The UA should use the computed values of the 'background-*', 'border-*', 'padding-*', 'outline-*', and 'text-decoration' properties when they do not have their initial values and the computed values of 'color', 'font', and 'cursor' (whatever their values) to influence the rendering where possible. Any values from those properties that cannot be used to influence the rendering of the UI control must not affect the rendering at all. For example, the UA should not draw a second border around a UI control that already has a border. If 'background-color' or 'background-image' have non-initial values and the UA is using their values for influencing the rendering of the UI control, then the UA must ensure that the 'color' property is also used to influence the rendering. 

That seems to have gotten lost along the way... //CC @tantek

**Comment by mfreed7** [comment]
> Is there a reason this can't be done via `color` alone? Would that currently break the modern OS appearance of the control?

I think the biggest downside to re-using `color` in this way is that it is non-specific. It would not allow you to, for example, change the text color to one thing, and the "accent" features to something else. Take, for example, the date picker:

```html
<input type=date style="color: red">
```

which has a text display of the date, and (on some platforms) a picker icon or arrow. If we tie everything to `color`, the above `<input>` would have **both** red text **and** a red icon. In contrast, `accent-color` would be specific to just the "accents", allowing black text with a red picker icon, for example:

```html
<input type=date style="color: black; accent-color: red;">
```

This would also be compatible with the existing Mac concept of an "Accent color", which affects control elements but **not** text colors. It would also be safe for web compat, since `accent-color` is a new CSS property.

**Comment by heycam** [comment]
I agree that tying this to `color` doesn't seem right.

A few questions that come to mind that I haven't given much thought to yet:

* On macOS, the system accent color affects how focus rings are rendered.  Should `accent-color` be allowed to influence focus styling?
* Is there any relation to `<meta name="theme-color">`?
* Existing form control designs may be amenable to some accent colors but not others.  For example, the Chrome checkbox design wouldn't work with `accent-color: white` unless the checkmark color changes in response.  Is this anything to worry about, or is it sufficient to have some wording that encourages engines to ensure their designs can adapt to all choices of `accent-color`?
* Do we need to think about how a future form control styling mechanism would interact with `accent-color`?  What about Web components that implement form controls, and which want to adapt their styling to the accent color?

**Comment by henke37** [comment]
An alternative would be to use pseudo elements to target the classic `color` property to specific parts of the form control. Naturally, this is not as generic as an accent color, but at the same time allows for detailed control of the color for individual parts.

This is not a mutually exclusive alternative, but interaction between the two would need to be clarified.

**Comment by chrishtr** [comment]
> * On macOS, the system accent color affects how focus rings are rendered.  Should `accent-color` be allowed to influence focus styling?

Focus rings can already be styled with CSS like  `:focus { outline: 5px solid orange }`.

> * Is there any relation to `<meta name="theme-color">`?

For some sites, yes, but I can also see them also wanting to choose a different color.

> * Existing form control designs may be amenable to some accent colors but not others.  For example, the Chrome checkbox design wouldn't work with `accent-color: white` unless the checkmark color changes in response.  Is this anything to worry about, or is it sufficient to have some wording that encourages engines to ensure their designs can adapt to all choices of `accent-color`?

I think we should add the language you suggested.

> * Do we need to think about how a future form control styling mechanism would interact with `accent-color`?  What about Web components that implement form controls, and which want to adapt their styling to the accent color?

We could perhaps define a pseudo element on custom form controls that accent-color automatically applies to? That's kind of what the `accent-color` I am proposing conceptually does.

**Comment by henke37** [comment]
"A" pseudo element. As in One. That has one name, suitable for any form control? X to doubt. If form controls get pseudo elements, there should be as many as that form element needs. Not some generic one size fits all one.

One pseudo class that applies to the correct pseudo elements could work tho.

**Comment by othermaciej** [comment]
What's the intended scope of `accent-color`? Is it just color of the checked state of radio buttons and checkboxes? If so, then the name seems inappropriately broad. If it's more things than that, then that may shed light on the name, and whether it should all be one property.

**Comment by chrishtr** [comment]
> Is it just color of the checked state of radio buttons and checkboxes?

Yes, something like that (or similar accents in other form elements).

**Comment by henke37** [comment]
"Something like that" and "or similar accents" are not phrases that belong in a standard. We need a clear definition of what "accent" means here, not vague hand waving.

**Comment by jensimmons** [comment]
How does this idea fit into the work of the Open UI project? @gregwhitworth ?

**Comment by gregwhitworth** [comment]
Windows also has the concept of an accent color on Windows 10. I'd recommend that this be a system color so that it can be leveraged in numerous places. As @chrishtr noted it's used in focus, but also can be used in the glyphs or other areas for form controls. I haven't looked at other use cases for this but the fact that Windows changes the title bars and other backgrounds to the accent color makes me think that it may be something that is allowed to be broad while allowing UAs to leverage it in our stylesheets for the specific scenario denoted by @chrishtr.

TLDR; I propose `accent-color` to be added to the [CSS System Colors](https://drafts.csswg.org/css-color-4/#css-system-colors) as a keyword

> How does this idea fit into the work of the Open UI project? @gregwhitworth ?

While this was discovered due to form controls it doesn't have to do with defining behavior, anatomy, states and interactions between states so I think staying in the CSSWG makes the most sense. Shoutout to @heycam for pinging the Open UI discourse channel though :)

**Comment by othermaciej** [comment]
@gregwhitworth I think the proposal here is to provide a way to _set_ the accent color for a particular form control, not to _use_ the system accent color (that may also separately be useful, but it also creates fingerprinting surface, which this proposal does not).

**Comment by mfreed7** [comment]
> What's the intended scope of `accent-color`? Is it just color of the checked state of radio buttons and checkboxes? If so, then the name seems inappropriately broad. If it's more things than that, then that may shed light on the name, and whether it should all be one property.

Ideally, the `accent-color` property can be specified to apply to all form controls, as appropriate. The appearance and styling of form controls is (deliberately) not well-specified, so it might be tough to avoid somewhat vague language here. However, I think the concept is reasonable. For example, in Safari, the system Accent Color setting affects:
- The "background" color for `<input type=checkbox>`
- The "background" color for `<input type=radio>`
- The activation "button" for `<select>`
- The activation "button" for `<input type=text list=datalist>`
- The highlighted row of `<select multiple>`
- The active state of `<button>`

All of the above seem to make sense as a usage of the accent-color. In Chrome, I could also see adding to the above list:
- The shaded portion of the `<range>` control
- The clear ('X') button on several controls such as `<input type=search>`

Again, the idea is to control the currently-uncontrollable colors of form elements just a bit better. Our hope is that the more broad Open UI type improvements in stylability will give even better control over all pieces of these controls. But in the meantime, this `accent-color` proposal will alleviate many very real, very [painful](https://bugs.chromium.org/p/chromium/issues/detail?id=1091531) developer problems with the existing control styling.

**Comment by heycam** [comment]
> The highlighted row of `<select multiple>`

On macOS Safari, this is rendered using the Highlight Color rather than the Accent Color. (By default the Highlight Color is a lighter or darker version of the Accent Color, but you can set it to a completely different color.)

**Comment by chrishtr** [comment]
> "Something like that" and "or similar accents" are not phrases that belong in a standard. We need a clear definition of what "accent" means here, not vague hand waving.

Yes, I'm aware of that. No need to talk down.

However, each user agent has their own default form control styling, so the specification of this property will need to higher level than specific DOM elements.

**Comment by othermaciej** [comment]
> However, each user agent has their own default form control styling, so the specification of this property will need to be higher level than specific DOM elements.

Going over how this might apply to rendering of a wider range of controls as in [this comment](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-652664697) might help us understand if all of this coming through one property makes sense. For example, Cameron noticed the [highlight color vs accent color issue](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-652672345).

It might help to be even more exhaustive in considering the full range of form controls. For example, one thing to consider is the `<input type=number>` stepper button. 

There are also colored pieces of form controls that do *not* normally follow the OS accent color. We probably want to avoid including those. Rationale: since accent color is likely to be desirable set across a whole site, that is the common use of similar functionality in apps. So it would be tempting to set it as a rule on `*`, but we wouldn't want unexpected parts of a form control to recolor.

**Comment by gregwhitworth** [comment]
@othermaciej ahhh my bad - completely mis-read the issue :/  I'll avoid the fingerprinting discussion given this is about a different proposal altogether.

> There are also colored pieces of form controls that do not normally follow the OS accent color. We probably want to avoid including those. Rationale: since accent color is likely to be desirable set across a whole site, that is the common use of similar functionality in apps. So it would be tempting to set it as a rule on *, but we wouldn't want unexpected parts of a form control to recolor.

LOL, I was just writing something similar. I agree that we should scope this in if it's solely focusing on form controls. Also, regarding the stepper buttons on `number` I think it's fine to still allow them on any replaced form control as in the future a UA may use the `accent-color` on the steppers or some other part that isn't currently exposed. Outside of that, if this is a common enough pain point now to add this to the platform then I am fine with this proposal as long as we scope and name it accordingly. How about `form-control-accent-color`?

**Comment by mfreed7** [comment]
Per the action items from the [CSSWG meeting](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-652700033), I have worked through these two action items:

1. Study/enumerate the various places where accent-color might apply, across form elements, platforms, and browsers.
2. Propose spec text for the ‘accent-color’ property.

Additionally, I gave some thought to the foreground/background color issue brought up by @frivoal. 

I've collected the results in this shared doc:

  https://docs.google.com/document/d/1yHQUshdC3hKrDRG-xDtUPYVIcc_JNUglK0KMVoqRyCo

I disabled comments on the doc, but I would love to hear feedback, so please add your comments back here in this issue thread.

**Comment by hwikyounglee** [comment]
> https://docs.google.com/document/d/1yHQUshdC3hKrDRG-xDtUPYVIcc_JNUglK0KMVoqRyCo

A clarification question:  Is there an assumption that there would be granularity options to apply the colors from all UA-chosen form controls to a single element chosen by the developer? Or would the granularity need to be just all UA-chosen controls?

**Comment by mfreed7** [comment]
> a clarification question for @mfreed7: Is there an assumption that there would be granularity options to apply the colors from all UA-chosen form controls to a single element chosen by the developer? Or would the granularity need to be just all UA-chosen controls?

Thanks for the question. This is a proposed new CSS property, which can be applied with full CSS granularity, down to individual elements. So:

```css
input[type=text].group1 {
  accent-color-foreground: blue;
}
#myelement {
  accent-color-background: red;
}
```

**Comment by chrishtr** [comment]
Agenda+ to discuss a resolution regarding the general shape of the spec approach Mason listed in his doc.

**Comment by heycam** [comment]
Regarding separate accent foreground and background colors, are there existing form control designs where the two colors can be independently chosen?  Or is it just that the accent foreground color is chosen automatically based on the background?  For example, on macOS you can select the accent background color from a specified set of colors, but the foreground color is chosen automatically (it's white, unless the highlight color is "Graphite" then it's black).  I'm not convinced we really need the two separate properties, as opposed to having requirements that, if they do honor accent-color, implementations must do so in a way that does not cause contrast issues.

**Comment by heycam** [comment]
Thinking about how that would apply to some of the examples listed in the document where current implementations don't use the accent color, but which a future design might like to: in Chrome, the activation widgets in some `input` elements are shown as a black icon on top of whatever background color the element has.  If an updated form control design were to take accent-color into account, then I could imagine the design being

* render a background color behind the icon, in the accent-color, and then automatically choosing black or white for the icon itself
* render a black or white background color behind the icon, and using accent-color for the icon itself

or something like that.  Of course there's a whole range of designs you could come up with, but I'm not certain that the accent being composed of two (hopefully contrasting!) author specified colors is necessary.

**Comment by gregwhitworth** [comment]
I went ahead and glanced at component libs across the web as well and here are the [results](https://docs.google.com/spreadsheets/d/16DS4Dt-I2NoX7pB9ZW6fLYOnTgZdHlUHvO7tM0N3ZvY/edit#gid=0).

Every single one that I can find allows some type of complete theme change. This is somewhat similar to `prefers-color-scheme` in the platform but it allows you to denote specific ones (primary, secondary, danger, etc). There are a few that allow you to pass in specific params or even select a custom color to override the where that color is used in the component (similar to accent color above) but only Material has the color helpers to ensure color contrast. Based on what I've seen from the component libraries and how the themes work you select a single color and the control adjusts accordingly while ensuring contrast. This aligns with what @heycam proposes with regards to specifying it in a way that the UA will do the color math based on the accent color as that is what the author is going to desire.

Now, designers commonly don't desire true black or white and so I can see them wanting to be able to control the foreground color. If it isn't a lot of additional impact on implementation I would specify something like for both foreground and background.

> If `accent-color-foreground` computes to `auto` and `accent-color-background` computes to a [CSS color](https://drafts.csswg.org/css-color-4/#typedef-color) then the user agent **MUST** use a color that meets the [WCAG minimum contrast requirements](https://www.w3.org/TR/WCAG/#contrast-minimum) of 4:5:1

This will achieve what @heycam denotes above while also allowing designers that want to still adhere to the WCAG but possibly in their own way do so.

**Comment by 1jj** [comment]
What about an _optional_ second value for background-color?
`accent-color: blue yellow;`

**Comment by mfreed7** [comment]
> What about an _optional_ second value for background-color?
> `accent-color: blue yellow;`

That's not a bad suggestion, though that would likely need to just be a shorthand property for:

```css
accent-color-foreground: blue;
accent-color-background: yellow;
```

(or perhaps the reverse?)

**Comment by frivoal** [comment]
It seems somewhat unlikely to me that we'd want these two things to cascade separately, so we might not need the longhands at all.

Also, we don't necessarily need to do this, but we could allow not just `<color> <color>?` but rather `<color>+` (or `<color>#`), and allow any number of colors. Then, as we specify in greater detail how various types of widgets use the colors, as is being suggested, we clarify which color goes where, and if some widgets (like range?) can use more than 2, then the author can supply that. Anyway, I don't think we should rush into that, but having everything under a single property makes this path easy if we later want to take it, more so that adding an increasing number of properties.

**Comment by heycam** [comment]
A concern I have about accepting multiple colors is that whether the colors work well together is going to be quite dependent on the platform/browser specific form control design.  Two colors may need to be contrasting in one form control design and not in another.

**Comment by frivoal** [comment]
> A concern I have about accepting multiple colors is that whether the colors work well together is going to be quite dependent on the platform/browser specific form control design. Two colors may need to be contrasting in one form control design and not in another.
> 
I thought that this was what we said we'd solve by being somewhat specific, control by control, about how the colors get used. I am little skeptical about how well that can work, but it seemed to me to be the direction the group was taking.

Either way, *if* we are going to take more than 1 color, I think a single multivalued property is better than multiple properties.

**Comment by mfreed7** [comment]
Thanks for the additional comments. After thinking about it, and given the comments, it does seem that this might be the cleanest, and most extensible proposal:

```css
accent-color: <color>+
```

The first provided color would be the "foreground" color, the second is the "background", and for some controls (`<range>`?) there might be a use for the 3rd+ color. Having foreground come before background seems logical to me for some reason, but we can also reverse that.

I am in the process of sketching out more-specific text, control by control, and I'll publish that here soon.

Side note: at the CSS weekly call, it was suggested that Google Docs aren't great for publishing documents like this (due to inaccessibility in some parts of the world, and risks of loss if Google goes under), so I've converted the doc I linked above to an MD file [here](https://github.com/mfreed7/accent-color/blob/master/study.md). There is still the risk that GitHub goes under, but then this comment will go with it.

**Comment by othermaciej** [comment]
Providing n colors seems even less likely to work well across a range of different control designs than providing 2.

**Comment by mfreed7** [comment]
Ok, as I [mentioned](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-665323663), I've been working on putting together a more "elaborate" description of this feature, with more details about each control. In the process, and through some very helpful input from @heycam, I ended up changing from "foreground" and "background" to "contrasting" and "primary", respectively.

Please see the new proposal, here:

https://github.com/mfreed7/accent-color/blob/master/proposal.md

The document contains proposed spec text, along with non-normative descriptions of each control. The idea here is that we can discuss and iterate on these descriptions ahead of time, avoiding the "first mover" problem discussed above. I've attempted to write this up in as complete a way as possible, because the concern before was that there would not be a way to properly specify this feature. Having said that, the above link **should be treated very much as a draft**, and I would love to hear your comments and feedback. Those can be left as comments in this issue thread, or as issues against my repo, whichever is easiest for you.

Please see the doc above for full details, but just for ease-of-digestion, here is a copy of the proposed spec text:

# Proposed Spec Text

<pre>
<b>Name</b>: ‘accent-color’
<b>Value</b>: &lt;color>+
<b>Initial</b>: 'auto'
<b>Applies to</b>: form control elements
<b>Inherited</b>: yes
<b>Percentages</b>: N/A
<b>Computed value</b>: computed color, see resolving color values
<b>Canonical order</b>: per grammar
<b>Animation type</b>: by computed value type

The ‘accent-color' CSS property sets the color of the “accent” parts or pieces of
form control elements. The first provided &lt;color> value is to be used for
"primary" accent elements. If a second &lt;color> value is provided, that color
should be used for "contrasting" accent elements.  The third and subsequent
colors are only used on some form control elements in some cases, for additional
"accent" parts other than "primary" or "contrasting".

If any color is not provided, or if 'auto' is provided for any color value, then
the UA should attempt to select an appropriate color which offers good contrast
and visibility when paired with remaining provided colors, if any. In selecting
'auto' colors, if the operating system provides an “Accent Color” or similar user
setting, the UA is encouraged to respect that setting as much as possible. The UA
may use a similar, though not identical, color in some cases, for example to
enhance contrast or accessibility.

In limited circumstances, it is permissable for user agents to render the accent
parts of some controls using different colors than those specified by
'accent-color', for example to maintain design guidelines or accessibility
constraints. In those cases, the rendered color should be influenced as much as
possible by the specified 'accent-color'. For example, the UA may wish to only
render the checkbox glyph in either white or black; in this case, the selection
of white or black should depend on the 'accent-color' value, e.g. using the
luminosity of the provided color.

Not all form elements contain “accent” parts, and not all user agents use the same
“accent” parts in exactly the same way for the same form control. However, the
intention is that if the same or similar accent parts exist on a given form
element, it should be associated with the "primary" or "contrasting" colors in the
same way across user agents. This is important to ensure good interop of the
'accent-color' property. For that reason, there is a table of form elements
provided below, which serves as guidance on the various accent parts for each
control. While the table is not normative, it is intended to provide some
alignment and uniformity of implementation across user agents.

The <b>text content</b> of form control elements is explicitly <b>not</b> included in the set
of “accent” parts, as text content is already controlled by the <a href="https://drafts.csswg.org/css-color/#the-color-property">'color'</a> property.
In addition, the <a href="https://drafts.csswg.org/css-backgrounds-3/#background-color">'background-color'</a> property is often used to control the
rendering for some background parts of form controls - those parts are similarly
not included in the set of "accent" parts that are subject to control via the
`accent-color` property.
</pre>

**Comment by frivoal** [comment]
Without diving into the specific of which color is used where, I feel you're on the right path in term of how this should be specified. One comment though. Earlier, you said:

> The first provided color would be the "foreground" color, the second is the "background"

It seems to me you actually took the opposite approach, and that this leads you in many example to use "auto" as the first color, since it wouldn't show up anyway in various controls which don't use a background color. Can you comment on this?

**Comment by mfreed7** [comment]
> Without diving into the specific of which color is used where, I feel you're on the right path in term of how this should be specified.

Glad you think so! I'd be really interested to hear specific per-control feedback, as I think this process (haggling over each one) might take some time, but it's important to the spec here.

> One comment though. Earlier, you said:
> 
> > The first provided color would be the "foreground" color, the second is the "background"
> 
> It seems to me you actually took the opposite approach, and that this leads you in many example to use "auto" as the first color, since it wouldn't show up anyway in various controls which don't use a background color. Can you comment on this?

Yes. This took some thought (and again, some good feedback from @heycam) to push me toward the current proposal. The basic insight was that the critical "color-matching" accent part should always be the first one. It would be really nice if an author could use `accent-color: blue` and then the most "visible" or "page-matching" elements would turn blue, while any other elements (e.g. glyphs) would automatically provide contrast. In most cases, the "primary" element turned out to be the element I had previously been calling "background". So I reversed the order, and renamed them to be a bit less specific to where it is (foreground or background) and more specific to whether it is the primary accent or just something that needs contrast.

As for the examples I provided, I just wanted to show a number of different scenarios. I don't think many of these would be typical, and as you can see from the color choices I made, I'm definitely not a designer. However, if you believe I got one (or many!) of the controls backwards, and you think the "primary/contrasting" elements should be swapped, that's great feedback to have.

**It would be great to get some developer input on this proposal**, in particular the ergonomics of "primary" vs. "contrasting" and the order of the two. It would also be good to get input on whether there is a need to provide the "contrasting" color at all, or if it would be good enough to just control the "primary" color and always leave the contrasting elements to the UA. Personally, I think we should be pushing to open access to all of the colors being used on each control, but there is overhead and complexity to doing that.

**Comment by henke37** [comment]
At this point I think it would be a good idea to have some live demos that use vendor specific code/custom controls to let people test out which parts use which color.

**Comment by mfreed7** [comment]
> At this point I think it would be a good idea to have some live demos that use vendor specific code/custom controls to let people test out which parts use which color.

That can certainly be a next step, and I'm happy to prototype this in Chromium, but I need some buy-in on the proposal from other implementers first. I do see the point that you'll want to experiment with it before knowing which way is "right" for each control.

**Comment by atjn** [comment]
> It would be great to get some developer input on this proposal, in particular the ergonomics of "primary" vs. "contrasting" and the order of the two.

Some feedback from a developer trying to create this design:

![image](https://user-images.githubusercontent.com/20535393/90445815-98424f00-e0e0-11ea-96d1-512032ea4548.png) ([link](https://timewarp.atjn.dk/))
...but with the clock icon also pink.

In my head, the clock icon is a `primary` color, the same way the text color is `primary` and the black background is `secondary`/`contrasting the primary`. By that logic, i should write `accent-color: currentColor`.

The current spec draft is the other way around, and calls the clock icon a `contrasting` color:
![image](https://user-images.githubusercontent.com/20535393/90446599-f02d8580-e0e1-11ea-906c-18690034c523.png)

In my head, that is wrong and clashes with the way i perceive a `primary` color. Then on the other hand, i completely agree with the classification of blue as the `primary` color for the progress slider:
![image](https://user-images.githubusercontent.com/20535393/90446825-59ad9400-e0e2-11ea-81f0-79ccab7a29fc.png)

Looking at the two examples above, the background color is sometimes the `primary` color and sometimes the `contrasting` color. In my head, the background should consistently be either `primary` or `contrasting`.

In other words, i think the examples from [your study](
https://github.com/mfreed7/accent-color/blob/master/study.md) generally makes more sense. You just need to rename `foreground` to `primary` and `background` to `contrasting`. Or not; `foreground` and `background` are just as, if not more descriptive to me.

>  It would also be good to get input on whether there is a need to provide the "contrasting" color at all, or if it would be good enough to just control the "primary" color and always leave the contrasting elements to the UA.

For this specific usecase, i just need primary. But i am certainly looking forward to having full control over checkboxes and radio buttons. Only having a primary color would be limiting to me in that case.

**Comment by richiksc** [comment]
The new macOS Big Sur has an example we could look to - they allow developers to set an "accent color" in native apps. https://developer.apple.com/design/human-interface-guidelines/macos/overview/whats-new-in-macos/#app-accent-colors

It seems that this accent color is applied across form controls, such as checkboxes, dropdowns, sliders, and textfields, but additionally used as the text selection highlight color and applied to selected elements from lists.

**Comment by mfreed7** [comment]
@atjn thanks for the detailed input!

> Some feedback from a developer trying to create this design in Chromium:
> ...but with the clock icon also pink.
> 
> In my head, the clock icon is a `primary` color, the same way the text color is `primary` and the black background is `secondary`/`contrasting the primary`. By that logic, i should write `accent-color: currentColor`.

I definitely see your point. I had trouble defining which element comes first/second here. And again my goal was to allow `accent-color: blue` (single color only) to make the most "visible" part of each control blue. With that goal in mind, my concern with the clock input, which (in **current implementations**) has just a glyph but no background "panel" behind it, is: what happens if new implementations would like to have a background color behind **just** the glyph? E.g. similar to some implementations of `<select>`:

<img width="83" alt="Screen Shot 2020-08-18 at 2 58 52 PM" src="https://user-images.githubusercontent.com/29559695/90570013-c61ca780-e163-11ea-90a6-3e8ad3112716.png">

In that case, it would seem that the "primary" color, which you would like to match with the rest of the site, would be that background fill, not the glyph, right? I.e. the blue in the `<select>` above, **not** the white glyph. The canonical example is the checkbox, where typically you would want `accent-color: blue` to give you this:

<img width="26" alt="Screen Shot 2020-08-18 at 4 17 05 PM" src="https://user-images.githubusercontent.com/29559695/90574649-4fd17280-e16e-11ea-9d30-fb7589716aaa.png">

where the **background** is blue, not the checkmark glyph. 

Again, these are all reasons why I think it is important to be able to specify **both** colors, **and** to have an agreed upon set of definitions for these controls - so you can always achieve your desired clock colors. As currently defined in the proposal, your [example](https://timewarp.atjn.dk/) could be perfectly achieved with `accent-color: #171717 #ff67bd`. Right?

> > It would also be good to get input on whether there is a need to provide the "contrasting" color at all, or if it would be good enough to just control the "primary" color and always leave the contrasting elements to the UA.
> 
> For this specific usecase, i just need primary. But i am certainly looking forward to having full control over checkboxes and radio buttons. Only having a primary color would be limiting to me in that case.

Thanks for the input here.

**Comment by gregwhitworth** [comment]
Awesome work on this Mason. Just a few thoughts.

> for example to maintain design guidelines

Isn't this a get out of jail free card?

> Glad you think so! I'd be really interested to hear specific per-control feedback, as I think this process (haggling over each one) might take some time, but it's important to the spec here.

I think this is a **fundamental** contract for this feature to be of value. _You_ note as such in your text above it.

> This is important to ensure good interop of the `accent-color` property. 

You then take this a step further in saying:

> Not all form elements contain “accent” parts, and not all user agents use the same “accent” parts in exactly the same way for the same form control. However, the intention is that if the same or similar accent parts exist on a given form element, it should be associated with the "primary" or "contrasting" colors in the same way across user agents.

This reads to me that you're saying that UAs, if they have the same parts, should then informally standardize through their 
 implementation of accent-color for those overlapping parts. Due to the above two statements I don't think that the definitions of these parts and the application of accent color should be non-normative.

I understand that defining and agreeing on every part is **not an easy** undertaking for form controls. However, I'd prefer for us to begin defining those parts and their associated accent colors. And as each control part definition lands in this spec it allows implementation of that control.

We can have an escape hatch since I know there will be arguments that a UA should be able to implement a part using the accent color provided the user in a way that isn't defined by the spec but that breaks the contract outlined above for interop purposes. We should have the normative definitions that are assigned to standardized parts be required to use the accent color defined *IF* `accent-color` does *not* compute to `auto`.

This will then allow an author to look at any UA and know that the glyph within the checkbox *will* get the contrasting color while the background will receive the primary color. Again, the author can opt into the native design by allowing this to compute to auto.

Ultimately, this discussion and the one in Open UI makes me feel like we need an agreed upon base UA stylesheet to solve the real pain points form controls present to authors while ensuring we don't stifle innovation for UAs.

**Comment by atjn** [comment]
@mfreed7 no problem, this spec is great work!
I would like to make it clear that i understand your approach to the design system, and if that's the way the final spec lands, that would be fine :)

> As currently defined in the proposal, your example could be perfectly achieved with accent-color: #171717 #ff67bd. Right?

That is correct, but in that case i am defining a color that does nothing. It is only there to fill the first space so i can set the second value. That doesn't seem like a robust solution to me. If the order was swtiched, i would be able to define only the necessary color:
`accent-color: #ff67bd`

i understand that this would mean you would have to define `accent-color: auto blue` in order to set the background of a checkbox to blue, but the difference is that `auto` actually refers to something that will be displayed on the page (the checkmark glyph), so it makes more sense to have to define that IMO.

I think most of these problems could also be solved if we just had direct access to the primary and contrasting colors:
`accent-color-primary: <color>`
`accent-color-contrasting: <color>`
`accent-color-extra: <color>+`   (or something like that)

So there is definetely and argument to be made for direct access to the values, but the way i subjectively see it, the system would be most elegant with just the shorthand, but the order flipped. My argument can be summed up into:

|                             | if `primary` is background     | if `primary` is glyph   | either version, using direct access to values     |
| :------------------------------------------------------------------------------------ | ---: | ---: | ---: |
| accepts single value for inputs with background as primary (checkbox, radio) | ✔️ | ❌ | ✔️ |
| accepts single value for inputs that don't have backgrounds (textarea, time)   | ❌ | ✔️ | ✔️ |
| does not force the use of filler colors that don't exist                                        | ❌ | ✔️ | ✔️ |


> [...] what happens if new implementations would like to have a background color behind just the glyph? [...]
> In that case, it would seem that the "primary" color, which you would like to match with the rest of the site, would be that background fill, not the glyph, right?

Yes, that is probably true for many websites, but in [the design i posted earlier](#issuecomment-675141914), the glyph would still be the primary accent color, and i would want the background to be fully transparent.
What ends up being the primary accent color is dependent on the design choices on every page, which is also why i find it more intuitive to call them `glyph`/`foreground` and `background` accent-colors, or to base the `primary` assignment on the fact that glyphs are usually always the `primary` color.

> Again, these are all reasons why I think it is important to be able to specify both colors, and to have an agreed upon set of definitions for these controls - so you can always achieve your desired clock colors.

I completely agree.

**Comment by gregwhitworth** [comment]
@mfreed7 is there a reason that we aren't using primary, secondary, tertiary for names? As we discussed on the call and @fantasai correctly pointed out:

> fantasai gave example of radio button, where Chrome 83 and Safari differ in whether the accent color is used for foreground or background, but both are using the accent color reasonably

This would allow the UAs to follow their design systems as they will have defined these colors. Also, I think that accent-color makes OS/UAs map it to their accent color that can be adjusted by the user and may propagate into an aspect of the control replacing the primary/secondary of their own brand for user benefit.

Ultimately, I'm fine with the above general consensus about not needing every part defined and which gets which color. My only other question then however is how can we test this? For this specific feature it seems that the definition of interoperability will be that it is implemented and the "spirit" is followed. Not sure how to make that work outside of render tree tests (since the parts may not be accessible early on) to determine if the accent color is leveraged.

Finally, I'd be curious if the wording is adjusted to ensure that spirit and provide a few concrete examples (as you've already done) if that won't be sufficient. I think that will address @hober concerns and if we can determine an agreed way of testing which will effectively determine spec compliance that will ease my concerns. (basically, I'm trying to avoid @mfreed7 from having to do even more research if it won't move us towards resolution).

**Comment by mfreed7** [comment]
So in the CSSWG [meeting](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-676546952) today, I heard several things:
 1. There is a desire to see more examples of current implementations of each form control, across browsers and platforms. This will help people see what the variety looks like today, and gauge how well (or poorly) the [proposal](https://github.com/mfreed7/accent-color/blob/master/proposal.md) will apply to those controls.
 2. There is a concern that this new property would provide too much "power". They would prefer that the UA should take perhaps only a single accent color, and should decide how (or whether) to use that color. More like "try to use purple if you can", rather than "use purple for the checkbox background, and white for the glyph".

On the first point, I'm happy to do that work. I will likely expand the proposal with a discussion/background section containing a few selected controls' appearance on multiple browsers and platforms. My short list includes: checkbox, radio, select, and date.

On the second point, I do understand where the concern comes from, but I disagree. We've already [heard](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-675141914) from developers on this thread that they have specific styling needs which would **require** some level of consistency among browsers. In the [example](https://timewarp.atjn.dk/) given there, the desire is to specifically control the "clock" glyph on a time control, and to make the rest of the control background one solid color. If this spec were "loose", then some UAs might decide to use `accent-color` to change the glyph color, and others might change the background behind the glyph. There would be no predictability, and developers would once again be forced to "roll their own" implementation. Or worse, they'll try to browser-sniff so that they can invert the colors to keep the appearance constant. For this to really provide value, I believe there needs to be some level of predictability.

Let me take a crack at expanding the examples, and see if that leads to any different conclusions. This might take me some time, but I'll check back here when it's complete.

**Comment by henke37** [comment]
I had an idea just now. Should there be support for named accent colors? As in, colors for a specific type of accent.

The first use case that comes to mind is an "error" color, to highlight that something went horribly wrong. Progress bars would be an existing example of this use case.

**Comment by henke37** [comment]
I had an idea just now. Should there be support for named accent colors? As in, colors for a specific type of accent.

The first use case that comes to mind is an "error" color, to highlight that something went horribly wrong. Progress bars would be an existing example of this use case.

**Comment by mfreed7** [comment]
Since it's our performance review time, I'm looking for ways to procrastinate. So I pulled together examples of Checkbox, Radio, Select, Date, and Button, across many different browsers and platforms, and added this as a [new section in the proposal](https://github.com/mfreed7/accent-color/blob/master/proposal.md#existing-control-examples).

My impression, after looking through them, is that the `accent-color` proposal would work in a fairly straightforward way for all of them. But please take a look and post your comments here.

**Comment by atjn** [comment]
Just to acknowledge [mfreed7's comment](#issuecomment-676736412).
If this spec does not give me full control over the form styling, and is reasonably consistent cross-browser, i will cry a lot, and then go back to using a library with custom controls where every aspect is easily styled with normal CSS.

I think that's a relatively common opinion.

**Comment by mfreed7** [comment]
I wanted to address two very good points made by @hober in the last CSSWG [meeting](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-676546952) (paraphrasing):
 1. What about the radio button? For example, current Chrome uses a larger central blue "dot" while most other browsers use a white "dot" on a blue (or other color) background. Shouldn't those both be `accent-color: blue`?
 2. What about controls that use gradients? What happens to those when `accent-color` is used?

For question 1, I would say that in the interest of interop, Chrome would likely remove the blue border, and color the outside (background) with the "primary" accent color, and the larger central "dot" with the "contrasting" accent color. I.e. `accent-color: blue` would effectively invert the default appearance of Chrome's radio button. And that's ok - the developer's intention there is to get all of the radio buttons to look (at least color-wise) the same. Yes, that's a different appearance than the full `auto` radio button, but when `accent-color` computes to something other than `auto`, the developer is indicating what they'd prefer.

Question 2 is also interesting. In the [new section](https://github.com/mfreed7/accent-color/blob/master/proposal.md#existing-control-examples) of the proposal, examples with gradients can be seen under the "Aqua" and "Snow Leopard" rows for many controls. (No modern implementations seem to use gradients, that I could find.) There are a number of ways to deal with that, I think. One is to just notice that the gradients go from darker blue to lighter blue, and if `accent-color: green` is specified, replace the gradient with a dark-green to light-green gradient. The other way, which I think I'd prefer, is that if `accent-color` computes to something other than `auto`, replace the gradient with a solid-color fill of the specified color. Again here, the developer is indicating that they'd like a particular color - let's respect it.

**Comment by BillGoldstein** [comment]
Thoughts as how to integrate with forced colors/contrast?

**Comment by fantasai** [comment]
@mfreed7 If the developer hasn't specified `appearance: none`, then I think it makes more sense to be trying to match the OS styling or whatever. We can lock down the exact painting of simple form controls with `appearance: none`, but we shouldn't disconnect form control styling from platform conventions just because the author wanted to indicate an accent color preference.

**Comment by gregwhitworth** [comment]
This was discussed at length, and one key thing that was brought up a few times by @hober @fantasai and @frivoal that I wanted to try and have a more meta discussion around leveraging `accent-color` to see if it can help align the conflicting needs of UAs and authors (all valid issues mind you). It outlines a proposal that @fantasai and @frivoal have discussed in general in the past and shared with me. This tries to get that down on paper (in spirit, not normative tex) and also how this will impact `accent-color` and enable all of the above use-cases presented by authors but ensure that UAs have the freedom for innovation across form factors/platforms.

Comments welcome: https://docs.google.com/document/d/1LEkDvTrhZuediHfyg6TSAQpPFTX0wMiaLueZhxjC2PQ/edit?usp=sharing

**Comment by AmeliaBR** [comment]
During yesterday's teleconference, there was a lot of discussion among group members about how important it was to have normative text about which parts of the form element should get the accent or contrasting color. There was a lot of concern (on the call & in this discussion) about whether authors would use the property at all without assurances of consistent appearance.

I threw up a quick Twitter poll to gauge author reactions. Thanks for those of you who retweeted, although caveat that my own followers have a larger than average share of web accessibility devs.

About 1 in 13 respondents said that current native HTML form elements are good enough for them. Of the remainder, half said they need more complete styling than just changing colors. And the rest were evenly divided between whether they would only use this property if it was consistent cross browser vs whether they were OK with variations to match native styling.

[The tweet and results](https://twitter.com/AmeliasBrain/status/1298656779274825728) were:

> If CSS added a property that allows you to tweak native form elements (checkboxes, select, etc.) to use one or more colours you specify, would you use that?
> 
> Or would you still need to fully customize?
> 
> Would it matter if results looked different between browsers?
> 
> - **46.5%** no, I need full styling
> - **22.9%** only if all browsers same
> - **22.9%** color-tweaked native = 💯
> - **7.6%** unstyled forms are fine
> 
> 419 votes · Final results

**Comment by gregwhitworth** [comment]
One thing I would like to note however, is that while I do think there are some longer term goals that we should continue to pursue with regards to the non-normative text for the assignments of the colors to the parts, that I think by it being non-normative addresses @hober concerns while also enabling a general guide that Mozilla raised as it results to interop concerns by not defining them. I think as further efforts proceed we may be able to circle back on this non-normative text but I think that @mfreed7 has done a good job of finding the correct balance at unblocking a pain point for developers today without limiting innovation of UAs.

**Comment by mfreed7** [comment]
Ok, at the last [teleconference](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-680996476), there was significant discussion about encouraging interoperability while not constraining innovation, and the inherent conflict therein. I also heard a suggestion from @astearns that it might help to include more prose explaining the motivation and intent for the spec, because it will aid in interpreting the specification. So I've gone ahead and done that - please see the [new section of the proposal](https://github.com/mfreed7/accent-color/blob/master/proposal.md#motivation-and-intent). I also copied it inline below.

Separately, I heard a recommendation that separate issues be opened for the sub-questions being debated here:
 1. Should we go forward with `accent-color`? This was [resolved](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-680996476) yesterday.
 2. Should we constrain browser vendor innovation of form control elements? I have not heard anyone argue that we should, so I'm going to just take it as a given that we shouldn't.
 3. Should interoperability be a goal for the `accent-color` spec? I [filed an issue](https://github.com/w3c/csswg-drafts/issues/5480) for this question.

While there are additional issues, I think <span>#</span>3 is a key question. A decision there will guide the rest of the process.

Here is a copy/paste from the new proposal section:

---

# Motivation and Intent

This section is non-normative.

In implementing `accent-color`, there are two competing/conflicting goals:
1. **Encourage interoperability** among browsers. It is important that developers be able to expect similar, if not the same, behavior across browsers.
2. **Encourage** (and do not constrain) browser vendors' **innovation** of form control elements.

The goal of interoperability pushes this solution towards a more strict specification of exactly how to use each `accent-color` value on each form control element. However, the goal of allowing innovation pushes this solution **away** from strict specifications for how each form control looks or acts. This specification attempts to provide a good middle ground, which maximizes interoperability while minimizing constraints to innovation.

The general methodology for achieving the above compromise is to examine the set of existing form control elements ([as of 2020](https://github.com/mfreed7/accent-color/blob/master/proposal.md#existing-control-examples-as-of-2020)), and agree on the *basics* of the way `accent-color` is applied to each of those existing controls. It is *explicitly* recognized that each browser provides different implementations of each form control, with their own look and feel. This spec does not try to eliminate those differences. However, it does try to provide some level of uniformity and interoperability, where commonalities exist.

This will require judgement when applying the recommendations to new or existing form controls. The goal would be to adhere to the *spirit* of this spec as much as possible. For example, if the guidance states that an `accent-color` value should apply to a particular accent part, and the browser implementation of that part uses a gradient fill rather than a single color, then there may be multiple ways to "use" the `accent-color` value to affect the gradient rendering of that part. The goal would be to match the guidance where it makes sense, and when possible. That might mean replacing the gradient with a solid fill, or it might mean changing the endpoint values of the gradient to match the `accent-color` value *in aggregate*. Or it might mean another behavior entirely. The point is that the guidance should be consulted and used as input in determining how to proceed.

Further, this spec is careful to not **require** exact conformance of each form control with the `accent-color` spec. It merely encourages browsers to follow the guidance for form controls elements that are "close enough" to the existing set of controls. For brand new paradigms, input surfaces, control types, etc., this spec does not attempt to limit innovation. If there are similar **parts** of these new-paradigm controls, then as much as possible/reasonable, those parts should be guided by this spec. But this should not be seen as any limitation on innovation.

To assist in characterizing "close enough" as it relates to the accent parts of form control elements, the [Existing Control Examples](https://github.com/mfreed7/accent-color/blob/master/proposal.md#existing-control-examples-as-of-2020) section of this document includes many examples of several control types, pulled from different browsers, operating systems, and time periods. Except where noted, each of the control examples within each group should be considered "close enough" to the group that the guidance for that group should apply.

**Comment by mfreed7** [comment]
I've just [closed](https://github.com/w3c/csswg-drafts/issues/5480#issuecomment-709542140) the "interoperability" issue that I opened, as it didn't help move the conversation forward as I had hoped. It actually seems to have somewhat derailed it, and that's likely my fault. The big conclusion I took from that discussion was that [my proposal](https://github.com/mfreed7/accent-color/blob/master/proposal.md) bites off too much, and most people seem to agree that `accent-color` should be less about "full control" of colors, and (as @samuelbradshaw [nicely puts it](https://github.com/w3c/csswg-drafts/issues/5480#issuecomment-704322971)) more about "dusting" the existing controls with a color. In other words, `accent-color` seems like a good fit for the "I want to change the theme and ensure contrast / readability, but don't necessarily need to control form UX so tightly", and **not** a good fit for "I want to make sure my form control looks exactly this way". Other technology is better for the latter. And in this context, interoperability can take the form of "achieves the intended purpose reliably across browsers" as opposed to "generates the same pixels across browsers".

As such, I think we should consider other proposals for the `accent-color` feature. In particular, @fantasai has made several such proposals (e.g. [1](https://github.com/w3c/csswg-drafts/issues/5544) and [2](https://github.com/fantasai/accent-color/blob/master/proposal.md)), and we should consider those and any others.

**Comment by chrishtr** [comment]
Agenda+ to discuss the specific proposals mentioned in Mason's second paragraph, or the version Brad suggested [here](https://github.com/w3c/csswg-drafts/issues/5544#issuecomment-703049815).

**Comment by fantasai** [comment]
# CSS 'accent-color' Proposed Text

Drafted by Mason Freed, 11 September 2020<br>
Revised by fantasai, 7 October 2020

<br>

As discussed on [CSSWG Issue 5187](https://github.com/w3c/csswg-drafts/issues/5187), and at the [July 1, 2020](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-652700033), [July 22, 2020](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-662570409), [August 19, 2020](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-676546952), and [August 26, 2020](https://github.com/w3c/csswg-drafts/issues/5187#issuecomment-680996476) CSSWG meetings,
there is a desire to expand the stylability of form control elements,
in particular by allowing the specification of the “accent color” for various elements.

This proposal is a result of those discussions.

## Motivation and Intent

In implementing `accent-color`, there are two competing/conflicting goals:
1. **Encourage interoperability** among browsers.
It is important that developers be able to expect similar, if not the same,
behavior across browsers,
and most critically, to get **usable** results on all browsers and platforms
now and in the future.
2. **Encourage** (and do not constrain) vendors’ **innovation**
of form control elements.

The goal of interoperability pushes this solution
towards a more strict specification of exactly how
to use each `accent-color` value on each form control element.
However, the goal of allowing innovation pushes this solution
**away** from strict specifications for how each form control looks or acts.
This specification attempts to provide a good middle ground,
maximizing necessary interoperability while minimizing constraints to innovation.

The general methodology for achieving the above compromise
is to examine the set of existing form control elements
([as of 2020](#existing-control-examples-as-of-2020)),
and agree on where and how an accent color is considered to have been applied
to each of those existing controls,
and define that those uses of color are replaced by the specified `accent-color`.
It is *explicitly* recognized that each browser
provides different implementations of each form control,
with their own look and feel.
This spec does not try to eliminate those differences.

This will require judgement when applying the recommendations
to new or existing form controls.
The goal is to adhere to the *spirit* of the feature as much as possible.
For example, if the specification requires
that an `accent-color` value should apply to a particular accent part,
and the browser implementation of that part
uses a gradient fill rather than a single color,
then there may be multiple ways to “use” the `accent-color` value
to affect the rendering of that part.
It might mean replacing the gradient with a solid fill,
or it might mean changing the endpoint values of the gradient
to match the `accent-color` value *in aggregate*.
Or it might mean another behavior entirely.
Non-normative examples and guidance are provided
as input to help the UA determine how to interpret `accent-color`
as appropriate to its own platform and implementation.

## Proposed Spec Text

```
<pre class=propdef>
Name: accent-color
Value: auto | <<color>>
Initial: ''auto''
Applies to: all elements
Inherited: yes
Percentages: N/A
Computed value: computed color
Canonical order: per grammar
Animation type: by computed value type
</pre>

User interface controls on any given platform are typically designed as a cohesive set,
under a single, cohesive visual style.
On many platforms (though not all),
the visual style includes the use of an <dfn>accent color</dfn>,
a typically bright color that contrasts markedly
with the more utilitarian background and foreground colors.
Not every control uses the [=accent color=] in every state,
but it is nonetheless a core part of the controls’ color scheme.

The 'accent-color' CSS property allows the author to specify the [=accent color=]
for user-interface controls represented by the element.
An <dfn for=accent-color value>auto</dfn> value represents a UA-chosen color,
which should match the [=accent color=] of the platform, if any.
A <dfn for=accent-color value><<color>></dfn> value
specifies that color as the desired [=accent color=].

The UA should use the 'accent-color'
to draw whichever parts of the form control
would have otherwise been styled with an [=accent color=].
The UA may adjust the luminance or brightness of the color
or make color substitutions in other parts of the control
(e.g. switching an overlaid glyph from using 'color' to using 'background-color'),
as necessary to maintain contrast for legibility of the control,
and may generate variations of the color for gradients etc.
and to match the control to platform conventions for the use of the accent color.

For example ...
```

# Examples

This section is non-normative.

## `<input type=checkbox>`

A checkbox is typically composed of a square,
potentially overlaid by a “checkmark” glyph representing the checked state.

Certain visual styles use an accent color for the checked state:
some of these use it for the background of the circle,
others for its foreground.
In both cases, 'accent-color' would be used to style
the affected parts of the control.
For visual styles that do not use an [=accent color=],
'accent-color' will have no effect.

The following are all examples of checkboxes rendered with a blue 'accent-color':

| Browser   | Platform  | Sample | Notes |
|-----------|-----------| :---:  |-------|
| Firefox 79| Win       | ![Checkbox](proposal_files/Firefox/checkbox.png) ![Checkbox Unchecked](proposal_files/Firefox/checkbox_empty.png) | No use of accent color |
| Chrome 81 | Win       | ![Checkbox](proposal_files/Chrome81/checkbox.png) ![Checkbox Unchecked](proposal_files/Chrome81/checkbox_empty.png) | No use of accent color |
| Chrome 83 | Win/Light | ![Checkbox](proposal_files/Chrome83/checkbox.png) ![Checkbox Unchecked](proposal_files/Chrome83/checkbox_empty.png) | Accent color as checked item “background” |
| Chrome 86 | Win/Dark  | ![Checkbox](proposal_files/Chrome86_dark/checkbox.png) ![Checkbox Unchecked](proposal_files/Chrome86_dark/checkbox_empty.png) | Accent color as checked item “background” |
| Safari 13 | Mac/Light | ![Checkbox](proposal_files/Safari/checkbox.png) ![Checkbox Unchecked](proposal_files/Safari/checkbox_empty.png) | Accent color as checked item “background” |
| Safari 13 | Mac/Dark  | ![Checkbox](proposal_files/Safari_dark/checkbox.png) ![Checkbox Unchecked](proposal_files/Safari_dark/checkbox_empty.png) | Accent as checked item “background”, and also used to generate highlight and shadow colors around the edges |
| Snow Leopard | Mac    | ![Checkbox](proposal_files/Snow_leopard/checkbox.png) ![Checkbox Unchecked](proposal_files/Snow_leopard/checkbox_empty.png) | Accent color used to generate checked “background” gradient |

Depending on the specified 'accent-color',
in order to preserve sufficient contrast
between the square and the checkmark
the UA might need to make some adjustments.
For example,
when applying 'accent-color' as the square’s “background”,
the UA could either adjust the accent color’s luminosity or brightness enough
to contrast with the color used to draw the “checkmark”
or change the color used for the “checkmark” to conrast better with the 'accent-color',
for example by switching from using the scheme’s foreground color to its background color
or vice versa.
If using 'accent-color' for the “checkmark” instead,
a different adjustment might be needed
since the relevant contrast would be between the 'accent-color'
and the square’s fill color.

### `<input type=radio>`

A radio button is typically composed of a circle,
potentially overlaid by a “dot” representing the checked state.

As with checkboxes,
certain visual styles use an accent color for the checked state:
some of these use it for the background of the circle,
others for its foreground.
In both cases, 'accent-color' would be used to style
the affected parts of the control.
For visual styles that do not use an [=accent color=],
'accent-color' will have no effect.

The following are all examples of radio buttons rendered with a blue 'accent-color':

| Browser   | Platform  | Sample | Notes
|-----------|-----------| :---:  |-------|
| Firefox 79| Win       | ![Radio](proposal_files/Firefox/radio.png) ![Radio Empty](proposal_files/Firefox/radio_empty.png) | No use of accent color |
| Chrome 81 | Win       | ![Radio](proposal_files/Chrome81/radio.png) ![Radio Empty](proposal_files/Chrome81/radio_empty.png) | No use of accent color |
| Chrome 83 | Win/Light | ![Radio](proposal_files/Chrome83/radio.png) ![Radio Empty](proposal_files/Chrome83/radio_empty.png) | Accent color as checked “foreground” |
| Chrome 86 | Win/Dark  | ![Radio](proposal_files/Chrome86_dark/radio.png) ![Radio Empty](proposal_files/Chrome86_dark/radio_empty.png) | Lightness-adjusted accent color as checked “foreground” |
| Safari 13 | Mac/Light | ![Radio](proposal_files/Safari/radio.png) ![Radio Empty](proposal_files/Safari/radio_empty.png) | Accent color as checked “foreground” |
| Safari 13 | Mac/Dark  | ![Radio](proposal_files/Safari_dark/radio.png) ![Radio Empty](proposal_files/Safari_dark/radio_empty.png) | Accent color as checked “foreground” |
| Snow Leopard | Mac    | ![Radio](proposal_files/Snow_leopard/radio.png) ![Radio Empty](proposal_files/Snow_leopard/radio_empty.png) | Accent color used to generate checked “background” gradient |

Depending on the specified 'accent-color',
in order to preserve sufficient contrast
between the circle and the dot
the UA might need to make some adjustments.
For example,
when applying 'accent-color' as the “background”,
the UA could either adjust the accent color’s luminosity or brightness enough
to contrast with the color used to draw the “dot”
or change the color used for the “dot” to contrast better with the 'accent-color',
for example by switching from using the scheme’s foreground color to its background color
or vice versa.
If using 'accent-color' for the “dot” instead,
a different adjustment might be needed
since the relevant contrast would be between the 'accent-color'
and the circle’s fill color.

## `<select>` / `<input type=text list=datalist>`

Drop-down and combo-box controls are typically displayed as
a text area containing the currently-selected value,
and an activation “widget” (such as an arrow button)
which is used to pop up the list of options.

Some visual styles use an accent color for the activation widget.
In these styles, 'accent-color' would be used to style the widget.

The following are all examples of drop-down controls rendered with a blue 'accent-color':

| Browser   | Platform  | Sample | Notes |
|-----------|-----------| :---:  |-------|
| Firefox 79| Win       | ![Select](proposal_files/Firefox/select.png) | No use of accent color |
| Chrome 81 | Win       | ![Select](proposal_files/Chrome81/select.png) | No use of accent color |
| Chrome 83 | Win/Light | ![Select](proposal_files/Chrome83/select.png) | No use of accent color |
| Chrome 86 | Win/Dark  | ![Select](proposal_files/Chrome86_dark/select.png) | No use of accent color |
| Safari 13 | Mac/Light | ![Select](proposal_files/Safari/select.png) | Accent color lightened and used to generate drop-down widget “background” gradient |
| Safari 13 | Mac/Dark  | ![Select](proposal_files/Safari_dark/select.png) | Accent color used to generate drop-down widget “background” gradient |
| Snow Leopard | Mac    | ![Select](proposal_files/Snow_leopard/select.png) | Accent color used to generate drop-down widget “background” gradient |

Depending on the specified 'accent-color',
in order to preserve sufficient contrast
within the activation widget
the UA might need to make some adjustments.
For example,
when applying 'accent-color' as a background,
the UA could either adjust the accent color’s luminosity or brightness enough
to contrast with the color used to draw the arrows
or change the color used for the arrows,
for example by switching from using the foreground color to the background color
or vice versa.
If using 'accent-color' for the arrows instead,
a different adjustment might be needed
since the relevant contrast would be between the 'accent-color'
and the widget’s background.

## `<select multiple>`

A multi-select is typically composed of
a rectangular area containing several lines of text,
each corresponding to an option.
Depending on the visual style,
selected options might be identified as highlighted items
or identified by some adjacent icon such as a checkmark.
In visual styles where the highlight is intended to match the [=accent color=],
the UA would use specified 'accent-color'.


## `<button>` / `<input type=submit|reset|button>`

A button does not typically have any [=accent colored=] pieces.


### `<input type=range>`

A range input many parts, and thus many potential accent parts.
One is the thumb, which is the part of the `<range>`
that the user can drag along the track.
Another is the track that the thumb slides along,
which is sometimes shaded differently on the “filled” side of the thumb,
the portion of the range, between 0 and `range.value`.
Whichever portion of the range input is typically colored by the [=accent color=]
would be likewise affected by 'accent-color'.

![Range](proposal_files/range.png)


### `<progress>`

A progress bar is typically composed of a shaded track,
within which a portion of the track is shaded in a different color
to indicate the value of the control.
Whichever portion of the progress bar is typically colored by the [=accent color=]
would be likewise affected by 'accent-color'.

![Progress](proposal_files/progress.png)


## `<input type=text|search|password|email|tel|url|...>`

Text-input controls are typically rendered as an empty box
containing the value of the input as text,
and do not have any parts highlighted by the [=accent color=].
However, sometimes additional controls are embedded,
such as a “clear value” button,
or a “show/hide password” button.
If the visual style uses its [=accent color=] to render any of these,
then they would be likewise affected by 'accent-color'.

![Text Input with Clear Button](proposal_files/search.png)


## `<input type=number|date|time|datetime-local|week|month>`

A number of more complex input controls are typically rendered as
a text field that displays the currently-selected value,
and an activation widget/button which is used to bring up a pop-up “picker”.
As with text-input controls,
if the visual style uses its [=accent color=] to render any part of the activation widget,
then they are likewise affected by 'accent-color'.
Additionally, if the visual style uses its [=accent color=] within the pop-up “picker”,
then those affected parts, too, can be affected by 'accent-color'.

![Clock Widget](proposal_files/clockface.png)

**Comment by samuelbradshaw** [comment]
Should accent color be limited to form control elements, or should it also be considered for other "traditionally blue" things like focus outline rings and text selection? (Should that question be opened as a new issue?)

**Comment by tabatkins** [comment]
New issue, please. (Tag it against `[css-ui-4]`)

**Comment by caraitto** [comment]
@chrishtr @mfreed7 
For a accent-color of "auto", could getComputedStyle() be used to get the OS color scheme (concern is that this could be used to fingerprint the user's OS)? Or would getComputedStyle(elem).getPropertyValue("accent-color") just return "auto" or ""?

**Comment by josepharhar** [comment]
> For a accent-color of "auto", could getComputedStyle() be used to get the OS color scheme (concern is that this could be used to fingerprint the user's OS)? Or would getComputedStyle(elem).getPropertyValue("accent-color") just return "auto" or ""?

I don't think we should return the OS color scheme here due to fingerprinting, like you said. We don't even use the OS color scheme for form control accent colors yet in chrome, we only use it for focus rings and selection color right now.

**Comment by henke37** [comment]
My concern there is that javascript may need the value in order to load appropriate assets or for other customizations too complex for css to handle.

**Comment by caraitto** [comment]
@josepharhar Could we make sure that the spec mentions that the OS color scheme shouldn't be returned?

**Comment by fantasai** [comment]
Spec has been published on /TR at https://www.w3.org/TR/css-ui-4/#widget-accent

Note: The computed value is explicitly defined as “the keyword auto or a computed color”.

**Comment by mangelozzi** [comment]
Well done guys this makes code much simpler (chopped out masses of pseduo ::before svg elements to simulate checkboxes, that had all sort of issues with validation), big thank you!

I have a suggestion. What if `accent-color` could be applied to `<select>` or `<option>` to finally be able to change blue background when hovering over an option:

(Please note the very nice green check box, because the color theme is green, but can't change the blue in the select box):
![image](https://user-images.githubusercontent.com/19764314/165466424-55c99e32-acfc-4ace-97cc-2e0b5a8a6c66.png)

**Comment by josepharhar** [comment]
`<select>` frequently uses platform-specific implementations for the popup that we can't style. I think that the upcoming `<selectmenu>` element will be easier to style.

**Comment by samuelbradshaw** [comment]
The spec for `accent-color` says that "whichever parts of the element’s control(s) would have otherwise been styled with an accent color" should "use the specified accent-color" – so, I think the blue you see in the drop-down menu is a bug (or an incomplete implementation) in the browser you're using.
https://www.w3.org/TR/css-ui-4/#widget-accent
